vkrunner/vr-script.c

```diff
@@ -1923,108 +1958,181 @@ process_token_replacements(struct load_state *data)
+        data->filename = filename;
+        data->line_num = 1;
+        data->script = vr_calloc(sizeof (struct vr_script));
+        data->current_stage = -1;
+        data->current_section = SECTION_NONE;
+        data->clear_depth = 1.0f;
 
-        data.script->filename = vr_strdup(filename);
-        data.script->framebuffer_format =
+        vr_pipeline_key_init(&data->current_key);
+
+        data->script->filename = vr_strdup(filename);
+        data->script->framebuffer_format =
                 vr_format_lookup_by_vk_format(VK_FORMAT_B8G8R8A8_UNORM);
-        assert(data.script->framebuffer_format != NULL);
+        assert(data->script->framebuffer_format != NULL);
 
-        vr_buffer_set_length(&data.extensions, sizeof (const char *));
-        memset(data.extensions.data, 0, data.extensions.length);
+        vr_buffer_set_length(&data->extensions, sizeof (const char *));
+        memset(data->extensions.data, 0, data->extensions.length);
 
         for (stage = 0; stage < VR_SCRIPT_N_STAGES; stage++)
-                vr_list_init(&data.script->stages[stage]);
+                vr_list_init(&data->script->stages[stage]);
+
+        return data;
+}
+
+static bool
+load_script_from_stream(struct load_state *data,
+                        FILE *f)
+{
+        bool res = true;
+
+        assert(data && "|data| must be a valid pointer.");
+
+        do {
+                if (!read_line(f, &data->line))
+                        break;
+
+                if (!process_token_replacements(data)) {
+                        res = false;
+                        break;
+                }
+
+                res = process_line(data);
+
+                data->line_num++;
+        } while (res);
+
+        if (res)
+                res = end_section(data);
+
+        return res;
+}
+
+static bool
+load_script_from_string(struct load_state *data,
+                        const char *string)
+{
+        bool res = true;
+
+        assert(data && "|data| must be a valid pointer.");
+        assert(string && "|string| must be a valid pointer.");
 
         do {
-                if (!read_line(f, &data.line))
+                string = read_line_from_string(string, &data->line);
+                if (data->line.length == 0)
                         break;
 
-                if (!process_token_replacements(&data)) {
+                if (!process_token_replacements(data)) {
                         res = false;
                         break;
                 }
 
-                res = process_line(&data);
+                res = process_line(data);
 
-                data.line_num++;
+                data->line_num++;
         } while (res);
 
         if (res)
-                res = end_section(&data);
+                res = end_section(data);
+
+        return res;
+}
+
+static struct vr_script *
+load_script_end(struct load_state *data)
+{
+        struct vr_script *script = data->script;
+
+        assert(data && "|data| must be a valid pointer.");
```

Comment:
If data was NULL it would have already crashed when dereferencing data->script above. I think these asserts aren’t really useful.

Reply:
Removed.